src/watchmaker/__init__.py

```diff
@@ -251,6 +251,25 @@ def __init__(self, arguments):
 
         self.config = self._get_config()
 
+        # Ensure a valid environment type was selected
+        if 'salt' in self.config:
```

Comment:
A couple of things here... 

1. I think we can flatten these nested if statements by using the dict.get method with a default value. I.e...

```
salt = self.config.get('salt', {'config': {}})['config']
valid_environments = [str(x).lower() for x in salt.get('valid_environments', [])]
env = str(salt.get('environment', '')).lower()
if env not in valid_environments:
    msg = (
        'Selected environment ({}) is not one of the valid'
        ' environment types: {}'.format(env, valid_environments)
    )
    self.log.critical(msg)
    raise InvalidValue(msg)
```

2. However, all these checks are an indication we're doing this in the wrong place. And I'm a little wary of putting worker-specific logic (salt) in the watchmaker `__init__` anyway... We can do it this way for now, but if you want to take a shot at putting the logic in the worker, that would be ideal. Right now the contract with the workers executes only a `install()` method. But, we can extend the contract to run a `before_install()` method where we can do this type of config validation.

That will require some teasing apart of the `worker_cadence` methods in `managers/workers.py`, as well as adding the new `before_install()` method to the base classes in `workers/yum.py` and `workers/salt.py`.

Reply:
This has been flattened and a before_install method has been added both to the individual workers and to the worker manager